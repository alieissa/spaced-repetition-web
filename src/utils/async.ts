/** @format */
/**
 * Credit goes to https://medium.com/@fillopeter/pattern-matching-with-typescript-done-right-94049ddd671c
 * For the pattern matching code
 */
export type Untriggered = {
  type: 'Untriggered'
}
export function Untriggered(): Untriggered {
  return {
    type: 'Untriggered',
  }
}

const t: Optional<Answer, 'easiness' | 'quality' | 'interval'> = {
  id: '',
  content: 'fsdfds',
}
export type Loading<V> = {
  type: 'Loading'
  value: V
}
export function Loading<V>(v: V): Loading<V> {
  return {
    type: 'Loading',
    value: v,
  }
}

export type Failure<E> = {
  type: 'Failure'
  value: E
}
export function Failure<E>(e: E): Failure<E> {
  return {
    type: 'Failure',
    value: e,
  }
}

export type Success<V> = {
  type: 'Success'
  value: V
}
export function Success<V>(v: V): Success<V> {
  return {
    type: 'Success',
    value: v,
  }
}

export type Async<L = unknown, F = unknown, S = unknown> =
  | Untriggered
  | Loading<L>
  | Failure<F>
  | Success<S>

export type AsyncStatus<R, E, A> = Async<R, E, A>['type']

export type AsyncMap<R, E, A, U> = {
  [K in AsyncStatus<R, E, A>]: U extends { type: K } ? U : never
}

export type AsyncStatusMap<R, E, A> = AsyncMap<R, E, A, Async<R, E, A>>
export type AsyncPattern<T, R, E, A> = {
  [K in keyof AsyncStatusMap<R, E, A>]: (a: AsyncStatusMap<R, E, A>[K]) => T
}

export function match<R, E, A>(
  e: Async<R, E, A>,
): <T>(pattern: AsyncPattern<T, R, E, A>) => T {
  return <T>(pattern: AsyncPattern<T, R, E, A>) => pattern[e.type](e as any)
}
