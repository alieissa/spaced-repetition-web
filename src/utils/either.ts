/** @format */

/**
 * Credit goes to https://medium.com/@fillopeter/pattern-matching-with-typescript-done-right-94049ddd671c
 * For the pattern matching code
 */
export type Right<V> = Readonly<{
  type: 'Right'
  value: V
}>
export function Right<V>(v: V): Right<V> {
  return {
    type: 'Right',
    value: v,
  }
}

export type Left<E = Error> = Readonly<{
  type: 'Left'
  value: E
}>
export function Left<E>(e: E): Left<E> {
  return {
    type: 'Left',
    value: e,
  }
}

export type Either<L, R> = Left<L> | Right<R>

export type EitherType<L, R> = Either<L, R>['type']
export type EitherMap<L, R, U> = {
  [K in EitherType<L, R>]: U extends { type: K } ? U : never
}
export type EitherTypeMap<L, R> = EitherMap<L, R, Either<L, R>>
export type EitherPattern<T, L, R> = {
  [K in keyof EitherTypeMap<L, R>]: (either: EitherTypeMap<L, R>[K]) => T
}

export function match<L, R>(
  e: Either<L, R>,
): <T>(pattern: EitherPattern<T, L, R>) => T {
  return <T>(pattern: EitherPattern<T, L, R>) => pattern[e.type](e as any)
}
