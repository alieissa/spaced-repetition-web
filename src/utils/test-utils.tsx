/** @format */

import { combineReducers, configureStore } from '@reduxjs/toolkit'
import { render } from '@testing-library/react'
import React from 'react'
import { Provider } from 'react-redux'
import { RouterProvider, createMemoryRouter } from 'react-router-dom'
import decksReducer from 'src/modules/decks/decks.reducer'

// As a basic setup, import your same slice reducers
import { loginReducer } from 'src/modules/login'
import { signupReducer } from 'src/modules/signup'
import { verificationReducer } from 'src/modules/verification'

// This type interface extends the default options for render from RTL, as well
// as allows the user to specify other things such as initialState, store.
// interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
//   preloadedState?: PreloadedState<RootState>
//   store?: AppStore
// }

const reducers = combineReducers({
  login: loginReducer,
  signup: signupReducer,
  verification: verificationReducer,
  decks: decksReducer,
})

const withRedux = (ui: React.ReactElement) => {
  const store = configureStore({ reducer: reducers })
  return <Provider store={store}>{ui}</Provider>
}
const withRouter = (ui: React.ReactElement) => {
  const router = createMemoryRouter([{ element: ui, path: '/' }], {
    initialEntries: ['/'],
    initialIndex: 1,
  })

  return <RouterProvider router={router} />
}
export function renderWithProviders(ui: React.ReactElement) {
  // Return an object with the store and all of RTL's query functions
  return render(withRedux(withRouter(ui)))
}

export const flushPromises = () =>
  new Promise((resolve) => setTimeout(resolve, 0))
