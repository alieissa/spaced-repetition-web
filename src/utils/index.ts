/** @format */

import * as async from './async'
import * as either from './either'

export { async }
export { either }
