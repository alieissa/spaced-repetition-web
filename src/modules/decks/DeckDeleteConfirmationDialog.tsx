import { useEffect } from 'react'
import { useParams } from 'react-router-dom'
import { Message } from 'semantic-ui-react'
import {
  SPButton,
  SPModal,
  SPModalActions,
  SPModalContent,
  SPModalHeader,
} from 'src/components'
import { async } from 'src/utils'
import { useDeleteDeck } from './decks.hooks'

type Props = {
  name: string
  open: boolean
  onCancel: VoidFunction
  onClose: VoidFunction
}
export default function DeckDeleteConfirmationDialog(props: Props) {
  const params = useParams()
  const [deleteStatus, deleteDeck] = useDeleteDeck(params.deckId!)

  useEffect(() => {
    if (deleteStatus.type !== 'Success') {
      return
    }

    // Same behaviour as close
    props.onClose()
  }, [deleteStatus.type])

  const handleConfirm = deleteDeck

  return (
    <SPModal
      data-testid="deck-delete-confirmation-dialog"
      size="tiny"
      open={props.open}
      onClose={props.onClose}
    >
      <SPModalHeader>Delete Deck</SPModalHeader>
      {async.match(deleteStatus)({
        Untriggered: () => null,
        Loading: () => null,
        Success: () => null,
        Failure: () => (
          <SPModalContent>
            <Message data-testid="deck-delete-error" />
          </SPModalContent>
        ),
      })}

      <SPModalContent style={{ alignText: 'center' }}>
        Are you sure you want to delete {props.name}?
      </SPModalContent>
      <SPModalActions>
        <SPButton onClick={props.onCancel}>Cancel</SPButton>
        <SPButton data-testid="deck-delete-confirm-btn" onClick={handleConfirm}>
          Confirm
        </SPButton>
      </SPModalActions>
    </SPModal>
  )
}
