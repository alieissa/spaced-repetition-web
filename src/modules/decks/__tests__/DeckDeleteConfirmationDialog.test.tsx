/** @format */

import { faker } from '@faker-js/faker'
import '@testing-library/jest-dom'
import { act, screen } from '@testing-library/react'
import { rest } from 'msw'
import { setupServer } from 'msw/lib/node'
import { flushPromises, renderModalWithProviders } from 'src/utils/test-utils'
import DeckDeleteConfirmationDialog from '../DeckDeleteConfirmationDialog'

const deckId = faker.string.uuid()
const deckDeleteUrl = `${process.env.REACT_APP_API_ENDPOINT}/decks/${deckId}/cards`

const server = setupServer()
beforeAll(() => server.listen())
afterEach(() => server.resetHandlers())
afterAll(() => server.close())

const mountComponent = () =>
  renderModalWithProviders(<DeckDeleteConfirmationDialog open={true} />, {
    initialEntries: [`/decks/${deckId}/cards/new`],
    path: 'decks/:deckId/cards/new',
  })

describe('DeckDeleteConfirmationDialog', () => {
  describe('view', () => {
    it('should render correctly', () => {
      const { asFragment } = mountComponent()

      expect(asFragment()).toMatchSnapshot()
    })
  })

  describe('interaction', () => {
    it('should display error when deck deletion fails', async () => {
      // Assemble
      server.use(
        rest.post(deckDeleteUrl, (__, res, ctx) => res(ctx.status(422))),
      )
      mountComponent()

      // Act
      const deckDeleteConfirmBtn = screen.getByTestId('deck-delete-confirm-btn')
      await act(() => deckDeleteConfirmBtn.click())
      await act(flushPromises)

      // Assert
      const cardCreateError = await screen.findByTestId('deck-delete-error')
      expect(cardCreateError).toBeInTheDocument()
    })
  })
})
