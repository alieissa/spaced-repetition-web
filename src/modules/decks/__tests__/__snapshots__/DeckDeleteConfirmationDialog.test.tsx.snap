// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeckDeleteConfirmationDialog view should render correctly 1`] = `
<DocumentFragment>
  <div
    class="ui inverted top aligned page modals dimmer transition visible active"
    style="display: flex;"
  >
    <div
      class="ui modal transition visible active"
      data-testid="deck-delete-confirmation-dialog"
      style="box-shadow: none; border: 1px solid; border-radius: 0;"
    >
      <div
        class="content"
      >
        <div>
          Delete
        </div>
      </div>
      <div
        class="actions"
      >
        <button
          class="ui button"
          style="border-radius: 0;"
        >
          Confirm
        </button>
        <button
          class="ui button"
          style="border-radius: 0;"
        >
          Cancel
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
